State: Initial migration to global selectors directory

aduth opened the pull request:
This pull request seeks to introduce a new Babel plugin, [`transform-imports`](https://www.npmjs.com/package/babel-plugin-transform-imports), to enable named importing of modules saved as files in the `state/selectors` directory. It's anticipated that this technique could also be applied to importing `lodash` modules.

This pull request moves `canCurrentUser` from `state/current-user/selectors.js` to `state/selectors/can-current-user.js` and updates all files making use of the selector.

**Reasoning:**
- https://github.com/Automattic/wp-calypso/pull/5954#issuecomment-225177211
- https://github.com/Automattic/wp-calypso/pull/5954#issuecomment-225469105
- p4TIVU-3B9-p2
- p4TIVU-3zj-p2

**Testing instructions:**

Ensure that all Mocha tests pass:

```
npm run test-client
```

Verify that no regressions occur on any screen making use of current user capabilities (domain credit notice, "manage settings" button on CPT list screen where post type unsupported but configurable - testimonials & portfolios).

/cc @mtias 

Test live: https://calypso.live/?branch=add/transform-wpcalypso-selectors

aduth commented:
cc @mcsf @johnHackworth

blowery commented:
The thing that feels odd here is that we're doing something tooling outside babel won't be able to follow. If there were an actual file at state/selectors that was re-exporting everything in the members of that folder, I'd feel better about it. That would also give us an easy path to true tree shaking when it becomes available.

gziolo commented:
Do we want to move only shared selectors to the selectors folder? Where can I find more details behind this idea? I would like to understand better what is our goal here, I only sense that it's related to the discussions triggered by React EU conference :)

mtias commented:
@gziolo the idea is to move all selectors, similar to how [action-types are all defined together](https://github.com/Automattic/wp-calypso/blob/master/client/state/action-types.js). Selectors operate on `state` since they can potentially grab pieces from different sub-trees. This change acknowledges that nature, helping with overall discoverability and function naming consistency.

ockham commented:
> the idea is to move all selectors, similar to how action-types are all defined together.

Are we sure about this? I have some reservations about this, since it will make modularization (think core+plugins structure) -- if ever envisioned -- of Calypso considerably harder. I've always thought that while it's true that some actions and selectors might require different parts of the state tree, there are some more basic ones that are rather confined, and which sort of map to more basic entities (maybe sort of defining what Calypso's core is?), and some higher-order/derived ones... 

(My case for modularization is a) our own Calypso spinoffs, and b) that adoption by the OSS/WP world hasn't been huge, and much like WP itself it might benefit from such a core+plugins.)

mtias commented:
> since it will make modularization (think core+plugins structure)

I don't think this should be an inherent problem — we should find a way to easily add a new selector from a plugin/extension, something we also need to figure out for action-types. We can still expose them via `state/selectors`. This would be handy if we have a section in devdocs to explore them, perhaps also with query selectors.

In a way it comes down to what are selectors consuming. If it's `state` then we should group them at that level. Perhaps some selectors, should they work with `state.subTree`, could remain declared within the reducer file, but I don't quite see the benefit.

aduth commented:
Sorry for pushing this up and running off for a weekend vacation without proper reasoning included. I've added a few references to the original comment, and @mtias nicely summarized reasons for the change.

I'm not sure I understand your modularization concern @ockham , but I'll add a few more points here that may help clarify its intent (some redundant/expanding on @mtias's points):

- We chose early on that selectors always accept the entire global state as its first argument, for the reason that selectors may operate on multiple subtrees. It could have been argued that many selectors did not need the entire state, but this decision was made largely for consistency, predictability, and flexibility of selectors (revised implementations). 
 - With this in mind, placing selectors in subtree folders seems contradictory to this decision.
- Developer experience is improved by...
 - Not having to spend effort deciding where to look to find existing or create new selectors
 - Fuzzy-search or folder scanning of directory to find all available selectors
 - Tests (and selector logic for that matter) are easier to locate than scrolling through a large file
- Code is improved by...
 - Single import statement for any selectors used
 - Potentially smaller bundles by only including the exact selectors used, not all selectors in a file (admittedly a non-concern with proper tree-shaking)

That said, part of the reason for putting this PR together was to spark discussion around the change in direction. If I've missed your point, could you clarify what you think makes our current approach more modular?

aduth commented:
>If there were an actual file at state/selectors that was re-exporting everything in the members of that folder, I'd feel better about it.

Yeah, at the very least, we should have an `index.js` that describes in a comment some of the magic that's happening here. Whether we start adding the exports to the file now... ? It's a bit more involved for the developer writing the selectors, but it may set the right expectations once we have proper tree-shaking, since at which point it'd be expected of them to add their selectors to `index.js` anyways.

aduth commented:
- Changed to a much less widely used selector to keep PR size down, `canCurrentUser` instead of `getSelectedSiteId` (96bb093)
- Updated "Our Approach to Data" docs with new recommendations on selector directory (da09340)
- Create an `index.js` with explanation and list of exported selectors (c31b83e)

gziolo commented:
Don't we want to upgrade to WebPack 2 to take advantage of its built-in tree shake feature? As I understood from [the latest JavaScript Air podcast about wepack](https://javascriptair.com/episodes/2016-06-08/), WebPack 2 is able to eliminate all unused exports when you use ES6 modules syntax! Related comment from the podcast:

> So webpack 2 now understands the ES6 module specification and so your imports and exports and things. And because it has a whole knowledge of your entire dependency tree, it can track all the dependencies, all the imports, all the exports, and know which of the exports are not being used. And so all that it does is like when it transpiles all of your import statements, because obviously those will not run in the browser, webpack resolves those, but when it transpiles those, it will exclude any exports from that transpilation process that are not being used elsewhere. So the code all still lives inside of that module, but like the actual export piece does not. And so, that way if that code is being used within that module it's still in use, and so uglyfication isn't gonna come through and dead code eliminate it.

Taking that into account a file named `client/state/selectors/index.js` would be the most important piece to make it work with WebPack 2 :)

aduth commented:
Correct @gziolo , eventually we'll be able to drop the Babel plugin and import directly from the `index.js` file. But Webpack 2 (stable) has been _coming soon_ for a very, very long time 😆

gziolo commented:
I now see that they are on 2.1.x beta without stable 2.0.x ;) I had an impression that they spent hour talking about production version :D

gziolo commented:
> Verify that no regressions occur on any screen making use of selected site Redux state.

BTW: This test scenario is no longer valid.

aduth commented:
Couple more changes:

- Enforce through testing that any file in `state/selectors` also export its selector in `index.js` (dc99878)
- Add README.md to `state/selectors` directory (78aa498)

>I had an impression that they spent an hour talking about production version :D

I was hoping there'd be some announcement of impending release during the podcast myself 😉 

>BTW: This test scenario is no longer valid.

Thanks, updated the original comment.

gziolo commented on client/state/selectors/test/index.js:
```diff
@@ -0,0 +1,53 @@
+		} );
+	} );
+
+	it( 'should export every selector file', ( done ) => {
```
Nice, you read my mind 👍

ockham commented:
@aduth @mtias Here's one concrete example: We've already factored out `i18n-calypso` into a repo/npm of its own. At some point in the future, we'd like to translate Theme Details sheets -- and since they are SSR'd, that also includes the server side. Currently, the user locale is stored at `i18n-calypso` module level, which obviously cannot work on the server -- it would persist across requests, while we want something contained per-request. This is quite reminiscent of our original motivation to adopt Redux instead of Flux, and I think using Redux state lends itself to this use case quite naturally, too.

But in that case, `i18n-calypso` would need some means to add its subtree to the global state tree, while its selectors/reducers/actions would clearly still be part of the module itself.

And I think a similar problem is conceivable with entire sections that are part of Calypso (or its forks) now and that we might factor into separate modules later.

mtias commented:
I'm not sure separating `i18n.calypso` was a good idea, to be honest. :)

Regarding how to incorporate these items from external modules / extensions, we would need systems for this in every area — components, sections, selectors/actions/reducers, app-components — that we need to figure out. We already have the notion of a "project" via config file, we could augment that to also include references to plugins that would be able to inject into the main app.

ockham commented:
> Regarding how to incorporate these items from external modules / extensions, we would need systems for this in every area — components, sections, selectors/actions/reducers, app-components — that we need to figure out. 

True. It's just that the Redux subtree sort of stood out to me, since -- in contrast to components and sections, which we strive to keep relatively contained -- we seem to assume that every selector and actions will possibly touch the entire Redux tree and that thus it's fine to just keep them all in one flat folder. It's the latter part I'm obviously somewhat skeptic about.

> We already have the notion of a "project" via config file, we could augment that to also include references to plugins that would be able to inject into the main app.

Yeah. I think with our rather contained components and sections, we could rely on npm/package.json for the dependency management, while possibly moving the section-local parts of our current `sections` or `project.js` to individual sections, and using the project-scope `project.js` to basically just enable sections (as [discussed earlier](https://github.com/Automattic/wp-calypso/pull/5487#issuecomment-221254741)).

blowery commented:
> Currently, the user locale is stored at i18n-calypso module level, which obviously cannot work on the server -- it would persist across requests, while we want something contained per-request.

That's a problem with `i18n-calypso`, not really a problem with having all of our selectors in one folder. Likely `i18n-calypso` needs to be refactored to ask for the state from a provider, which the user of i18n-calypso has to provide. That could be backed by redux / express context / whatever.

ockham commented:
> That's a problem with `i18n-calypso`, not really a problem with having all of our selectors in one folder. Likely `i18n-calypso` needs to be refactored to ask for the state from a provider, which the user of i18n-calypso has to provide. That could be backed by redux / express context / whatever.

Yeah, guess that's right. What about the sections plugins case?

ockham commented:
(Anyway, I don't want to de-rail this PR -- especially if the general sentiment is in favor of such a solution, and while my use-cases are still highly theoretical.)

nylen commented:
I echo @blowery's concerns on this change - I'm not a big fan of the idea to use tooling that makes our code further from standard JS.  If we continue down this path then we're writing "CalypsoScript" rather than JavaScript/JSX.  I don't think this is a good thing for code reuse and interoperability with other projects.

If we want to change the import syntax for selectors, we should create a plain js file that contains the needed export statements, whether this is done by hand or by an automated process.

aduth commented:
>I'm not a big fan of the idea to use tooling that makes our code further from standard JS.

There's nothing non-standard about what's going on here. We could rip out `babel-plugin-transform-react-display-name` from this pull request and it'd work just as well. The difference is in some magic to prevent the bundles from becoming unnecessarily large because of the imports that are dragged in from the `client/state/selectors/index.js` file (Webpack 2 and [Rollup](https://github.com/rollup/rollup) use "tree shaking" to achieve a similar effect).

nylen commented:
👍 I see now that as of https://github.com/Automattic/wp-calypso/pull/5954/commits/c31b83eaaf6620535515c8f92d3fa662e5c4896b the PR includes an index.js which behaves as expected.

mtias commented:
@aduth how are you feeling with this one?

aduth commented:
I still generally agree with the idea behind it and have been taking this approach in a side project of mine. The one-per-file selector feels a bit... excessive at times, but I do like the option to `import from 'state/selectors';`. I have a hard time not using the same arguments to conclude that action creators are similarly global and should also live separate from the state shape files, which deters me from wanting to push this forward, as I like some semblance of file organization.

mcsf commented:
> I have a hard time not using the same arguments to conclude that action creators are similarly global

I remember a rather philosophical convo with Matías last year getting to this very conclusion.

> which deters me from wanting to push this forward, as I like some semblance of file organization

I understand the hesitation, and TBH don't know how to weigh in right now.

jordwest commented:
Would it make sense to have both global and local/modular selectors?

For example, the `isOlarkReady` selector only needs access to `state.ui.olark.*`, so it makes sense to be located in `/state/ui/olark/selectors.js`. On the other hand, if it needed access to `state.ui.editor.*`, it might be better placed in `/state/ui/selectors.js`.

Finally, if it needed access to `state.site-settings.*` **and** `state.ui.olark.*`, then `/state/selectors.js` would be the natural home for it.

aduth commented:
Your comment does pretty well to highlight why a global selector directory might be the ideal approach, since when using a selector, I shouldn't need to concern myself with the areas of the state that selector needs. Especially when considering that could change over time. On the flip-side, there's more overhead in having a global selector directory, as creating individual files per selector can be a bit tedious.

gziolo commented:
What do we need to move forward with this one? :)
I'm happy to do a final check if we decide to go this route.

aduth commented:
@gziolo : Thanks for the nudge 😆  I'd lost my interest in championing this, though I still think it's a problem that needs to be solved. As I see it, the general issue is discoverability, both in terms of knowing which selectors are available and where to find them.

Summarizing a few of the worries with a global selectors directory:

- Cumbersome / verbose to create a single file per selector
- What do we do with the miscellaneous files with utilities used by selectors?
  - Example: [`state/posts/utils.js`](https://github.com/Automattic/wp-calypso/blob/master/client/state/posts/utils.js)
  - It's admittedly a pretty gross convention, as it has the tendency to become a dumping ground, but these are common functions used across multiple selectors
- Loses benefits of scalability from folder organization, even if the selectors don't always fit cleanly within those buckets
- Once we start defining selectors as global, how do we proceed with actions?

And the benefits:

- Selectors discoverable in that they're located in the same directory
- Selector dependencies are much more obvious
- Test files are easier to navigate
- Accurately reflects selectors as being able to traverse multiple state subtrees
- Importing selectors is easy and concise (`import from 'state/selectors';`)

These are summaries of the situation, and I personally don't see a clear conclusion. I've tried to think of other ways we could solve the original issue, and would be curious to hear people's thoughts on whether either of the following ideas are worth pursuing:

- DevDocs screen for listing and/or searching selectors
  - Maybe enhanced with details from our meticulously crafted JSDoc blocks
- Webpack loader which "discovers" selectors by our conventions and allows for `import from 'state/selectors';` without files necessarily being located there

The first seems like something we should probably do anyways. The second smells a bit magical, and only solves the "where to find them" discoverability problem, not "which selectors exist".

When it's all said and done, I do think the changes here are an improvement over our current situation, and if we're unable to come to an agreement on an alternative, I'd be happy to revive this aging pull request.

lancewillett commented:
@gziolo @aduth This came up in a scrub for older, stale pull requests. Still pending review, or is it in-progress or stalled?

aduth commented:
@lancewillett I've had renewed interest in this pull request over the past week. I'll try to breathe life back into it again this week.

aduth commented:
Rebased, squashed, and improved some of the documentation in the changes here to revive this effort. It was interesting to see how widespread selectors like `canCurrentUser` have become (previous to rebase had only existed in two files, now 12).

I had lost interest in this previously because after trying this approach in a side project, I worried about the overhead of having many individual files in a `state/selectors` directory. However, in the past couple months, it's become obvious that there's a big issue with discoverability of selectors and where is most sensible for a particular selector to exist. I think having a single selectors directory simplifies both of these questions, and opens up the possibility of better tooling around searching selectors. Per-file selectors also helps avoid potential merge conflicts which can currently occur when editing larger selector files, and can help reduce bundle sizes when only using one or few selectors currently located in a larger selectors file.

mtias commented:
@aduth Seems you need to rebase again. :) Ping me when you are done so we can merge it before it gets out of sync.

aduth commented:
Rebased. Conflict was in shrinkwrap, plus two more new usages of `canCurrentUser` accounted for.

mtias commented:
I think we should 🚢

aduth commented:
I plan to merge this early next week, hopefully after having spent some time exploring codemods for transitioning our legacy approach and DevDocs tooling for exploring selectors within this new directory. Both of these in the hopes of enticing adoption.

aduth commented:
I've put together a [jscodeshift](https://github.com/facebook/jscodeshift) codemod script that should help future migrations of imports to using the global selector directory:

https://gist.github.com/aduth/904f56aa5594af171d004dd8276b281e

Example usage:

```
npm i -g jscodeshift
jscodeshift client --selector=getSelectedSite --extensions js,jsx -t https://cdn.rawgit.com/aduth/904f56aa5594af171d004dd8276b281e/raw/18e17ee8f734b57501f09164c4c68b617fdb2388/migrate-global-selector.js
```

...replacing `getSelectedSite` with the name of the selector to be migrated.

Example output (single file):

```diff
diff --git a/client/blocks/url-preview/index.jsx b/client/blocks/url-preview/index.jsx
index 32eec51..fe16dc1 100644
--- a/client/blocks/url-preview/index.jsx
+++ b/client/blocks/url-preview/index.jsx
@@ -9,7 +9,8 @@ import debugFactory from 'debug';
  * Internal dependencies
  */
 import { closePreview } from 'state/ui/preview/actions';
-import { getSelectedSite, getSelectedSiteId } from 'state/ui/selectors';
+import {getSelectedSiteId} from 'state/ui/selectors';
+import {getSelectedSite} from 'state/selectors';
 import { getPreviewUrl } from 'state/ui/preview/selectors';
 import { getSiteOption, getSiteSlug } from 'state/sites/selectors';
 import addQueryArgs from 'lib/route/add-query-args';
```

Currently the spacing of imports is mangled but fortunately (and with convenient timing!) a configuration option is being added to the underlying library in https://github.com/benjamn/recast/pull/341 . The script already includes the configuration option so it should "just work" once a new version of `jscodeshift` is made available.

aduth commented:
I've rebased and will plan to merge this shortly.

Aside: The codemod in my previous comment works a treat and was useful in rebasing with more recent usage of `canCurrentUser`.

Next, I'll merge #9433 and put together a write-up for expectations and future plans (including, perhaps, a codemod to migrate all selectors in one fell swoop!)
